addons/bitbucket/views.py

```diff
@@ -0,0 +1,175 @@
+# -*- coding: utf-8 -*-
+import httplib as http
+import logging
+
+from flask import request, make_response
+
+from framework.exceptions import HTTPError
+
+from addons.base import generic_views
+from addons.bitbucket.api import BitbucketClient
+from addons.bitbucket.apps import bitbucket_hgrid_data
+from addons.bitbucket.serializer import BitbucketSerializer
+
+from website.project.decorators import (
+    must_have_addon, must_be_addon_authorizer,
+    must_have_permission, must_not_be_registration,
+    must_be_contributor_or_public
+)
+
+logger = logging.getLogger(__name__)
+
+
+SHORT_NAME = 'bitbucket'
+FULL_NAME = 'Bitbucket'
+
+############
+# Generics #
+############
+
+bitbucket_account_list = generic_views.account_list(
+    SHORT_NAME,
+    BitbucketSerializer
+)
+
+bitbucket_import_auth = generic_views.import_auth(
+    SHORT_NAME,
+    BitbucketSerializer
+)
+
+def _get_folders(node_addon, folder_id):
+    pass
+
+bitbucket_folder_list = generic_views.folder_list(
+    SHORT_NAME,
+    FULL_NAME,
+    _get_folders
+)
+
+bitbucket_get_config = generic_views.get_config(
+    SHORT_NAME,
+    BitbucketSerializer
+)
+
+bitbucket_deauthorize_node = generic_views.deauthorize_node(
+    SHORT_NAME
+)
+
+
+#################
+# Special Cased #
+#################
+
+@must_not_be_registration
+@must_have_addon(SHORT_NAME, 'user')
+@must_have_addon(SHORT_NAME, 'node')
+@must_be_addon_authorizer(SHORT_NAME)
+@must_have_permission('write')
+def bitbucket_set_config(auth, **kwargs):
+    node_settings = kwargs.get('node_addon', None)
+    node = kwargs.get('node', None)
+    user_settings = kwargs.get('user_addon', None)
+
+    try:
+        if not node:
+            node = node_settings.owner
+        if not user_settings:
+            user_settings = node_settings.user_settings
+    except AttributeError:
+        raise HTTPError(http.BAD_REQUEST)
+
+    # Parse request
+    bitbucket_user_name = request.json.get('bitbucket_user', '')
+    bitbucket_repo_name = request.json.get('bitbucket_repo', '')
+
+    if not bitbucket_user_name or not bitbucket_repo_name:
+        raise HTTPError(http.BAD_REQUEST)
+
+    # Verify that repo exists and that user can access
+    connection = BitbucketClient(access_token=node_settings.external_account.oauth_key)
+    repo = connection.repo(bitbucket_user_name, bitbucket_repo_name)
+    if repo is None:
+        if user_settings:
+            message = (
+                'Cannot access repo. Either the repo does not exist '
+                'or your account does not have permission to view it.'
+            )
+        else:
+            message = (
+                'Cannot access repo.'
+            )
+        return {'message': message}, http.BAD_REQUEST
+
+    changed = (
+        bitbucket_user_name != node_settings.user or
+        bitbucket_repo_name != node_settings.repo
+    )
+
+    # Update hooks
+    if changed:
+
+        # # Delete existing hook, if any
+        # node_settings.delete_hook()
```

Comment:
Note: Bitbucket does support webhooks, but Services/Product teams opted to skip them in the short term to release this faster. Comments left to simplify adding in future